### DISCUSSION TITLE
Evaluation time of declared_attr

### DISCUSSION THREAD

**Discussion by astoff** [sqlalchemy/sqlalchemy#11080] [Usage Questions] [answered]
I would like to store some [aliased](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.aliased) ORM entities as table attributes, generating them from a Mixin. However, code of this shape produces an error (`ArgumentError: Column expression, FROM clause, or other columns clause element expected`):

```python
class MyMixin:
    @declared_attr
    def aliased_thing(cls):
        subquery = select(cls).where(...).subquery()
        return aliased(cls, subquery)
```

The reason seems to be the timing of the evaluation of the `declared_attr`. For example, the code below gives an `AttributeError: type object 'X' has no attribute '__table__'`.

```python
class X(Base):
    @declared_attr
    def x(cls):
        print(cls.__table__)
```

Is there any way around this issue?

**Comment by zzzeek** [comment, marked as answer]
declared_attr functions can't work with the class as a mapped class because it's not mapped yet.

we dont have a dedicated "deferred evaluation" feature for column_property() as of yet so you would need to use a hook like `__declare_first__()` or `__declare_last__()` to automatically set things up on the soon-to-be-mapped class.   https://docs.sqlalchemy.org/en/20/orm/declarative_config.html#other-declarative-mapping-directives
